### ISSUE TITLE
Screen doesn't update when rotated/reflected

### ISSUE THREAD

**Issue by ApeironTsuka** [DisplayLink/evdi#343]
* Are you using the latest driver?
    Yes
  * Are you using the latest EVDI version?
    Yes; tested a fresh pull of this git, but it also occurred in 1.9.1
  * If you are using a DisplayLink device, have you checked 'troubleshooting'
    on DisplayLink's website?
    Yes
  * Is this issue related to evdi/kernel?
    I think so?

  * Linux distribution and its version: Debian Bookworm/testing
  * Linux kernel version: 5.15.0-2-amd64 (5.15.5-2)
  * Xorg version (if used): 1.20.13
  * Desktop environment in use: Xfce 4.16
  * NVidia proprietary blob driver

When any sort of transform is applied (rotated, reflected, etc), the screen's buffer is no longer updated/refreshed live except when the cursor is moved on the connected display. This can be easily demonstrated by running glxgears on the DisplayLink display, but it isn't limited to OpenGL-based applications.

On 1.9.1, this meant that rotating the screen to the right would cause the screen to "freeze" and only redraw the screen when the cursor was moved.
On 1.10.0, the screen only redraws the rectangle of the cursor itself when moved and nothing else.

On both versions, the screen updates fine with no transforms applied.

I unfortunately don't know enough about how the rendering/drawing pipeline works to dig into this myself, and don't really know where to start for learning about it.

**Comment by displaylink-emajewsk** [comment]
Hi, sorry for the wait.

Would you mind trying the newest driver and seeing what effect it has on the issue?

Also, if possible, I'd like to know more about your setup (mostly the docking station) and whether changing between x11 and wayland affects the issue.

**Comment by ApeironTsuka** [comment]
Sorry about that. Meant to do this earlier but it's been a long week.

Tested driver 5.5 from the site as well as using a fresh pull of this repo.

DisplayLink screen set to be right-of main panel. Upon applying a transform, the screen turns black and redraws the full screen only when the cursor is moved while on the DisplayLink screen. Same test as above, using glxgears as a simple redrawing window.

- The DisplayLink device is an ASUS MB169B+ panel.
- Kernel 5.16.0-4-amd64 (5.16.12).
- nvidia proprietary driver version 390.147 (older laptop, this is the nvidia-legacy-390xx driver).
- Xorg version 1.21.1.3

**Comment by kevinpthorne** [comment]
Currently experiencing the same issues, rendering the monitor completely useless when in a non-landscape mode

**evdi version**: 1.10.1-2
**displaylink version**: 5.5-1
both built from source

-----
**Linux kernel:** 5.15.41-1-MANJARO
**Xorg version:** 1.21.1.3
**Desktop env:** gdm 42 (Gnome)
**Nvidia driver version (proprietary):** 510.73.05

**Comment by microhobby** [comment]
Same behavior here 😥

**Comment by microhobby** [comment]
@displaylink-emajewsk could you check with the support team please? I'm almost sure that there is no workaround (or use it in the landscape only mode). Whenever I have to use Linux, I have to change my entire setup to accommodate the third screen in landscape. I'm considering looking for another USB video card vendor.

**Comment by b1g7r33** [comment]
So I found a bit of a workaround by accident  in Xubuntu 22.04. **TL:DR below.**

**Background:** I have and Ultrawide (3440x1440) and two 1080p monitors. The 1080p monitors are rotated in portrait, left and right of the Ultrawide. The respective 'top' of each monitor faces out from the center (not sure if it matters, not redoing cable management to find out). The two 1080p monitors are fed from a Displaylink dock. When I put any application on the 1080p monitors the application would only refresh when I was moving my mouse. Sound would play but the application view stays in a frozen state until I move my mouse. Almost like the window is off of the edge of the display not on a screen (this matters later).

**FIX:** I halfway fixed this through the 'Settings' application going through Window Manager Tweaks->Compositor tab->unchecking 'Enable display compositing'. No restart or Logout, it worked live. This fixes the rotation bug UNLESS the window is hanging off of the outer edge of the display space. 

**EXAMPLE:** So on the right monitor; if the application is hanging off of the right edge (the monitor's respective 'top' in my case) the application still doesn't draw video but audio plays. If it was hanging off of the left edge (lapping over onto the Ultrawide in the center) the application performs normally.

**TL;DR** Turn off display compositing in settings as a **TEMPORARY WORKAROUND** until Displaylink fixes this in a decade.

**Comment by sprietNathanael** [comment]
I'm having the same issue :

Linux kernel: 6.8.0-31-generic
Xorg version: 2:21.1.12-1ubuntu1
Linux distribution: Kubuntu 24.04
Desktop: plasmashell 5.27.11
Nvidia driver version (proprietary): 535.171.04
GPU: Nvidia GeForce GTX 1650

When I sitch the nvidia PRIME profile to "on demand", everything works fine. But when I switch it to "performance mode", my rotated screen does not refresh unless the cursor moves.

**Comment by FuzzyExpress** [comment]
I've also got this issue, Screen was great fine on my last laptop, but my new one has a iGPU that is incapable of doing On Demand without lag. I can confirm that refreshing works fine in horizontal orientation, however disabling the compositor didn't for it for me. 

Machine: Kubuntu Focus M2 Gen 5    (X370SNx1)
OS: Kubuntu 24.04.1 LTS x86_64
Kernel: 6.8.0-45-kfocus   (I do not have a normal kernel)
DE: Plasma 5.27.11
CPU: Intel i9-14900HX (32) @ 5.600GHz  
GPU: NVIDIA GeForce RTX 4090 Laptop GPU 
Nvidia driver version: 535.183.01
X.Org X Server 1.21.1.11
Driver version: 1.14.7
EVDI service version: 1.14.7

https://github.com/user-attachments/assets/0e1f168f-d14b-41f8-b0d3-e650c6921a35

Note: DisplayLink installed from [displaylink-debian](https://github.com/AdnanHodzic/displaylink-debian), the [official install method](https://www.synaptics.com/products/displaylink-graphics/downloads) couldn't reach 1 FPS and was unusable.

**Comment by blackPantherOS** [comment]
This problem has been occurring for a long time: cannot refreshing portrait orientation screen , landscape is okay

Tested with all Nvidia driver
Tested with all Kernel from 5.x -> 6.13.1
Evdi: : 1.14.8
Plasma5

**Comment by WojtekPachowiak** [comment]
Still an issue on my vertical monitor. My workaround is to connect the vertical monitor directly to ma laptop, rather than to my docking station.

OS: CachyOS x86_64
Host: HP EliteBook 830 G6
Kernel: Linux 7.1.3-2-cachyos
Display (EV2450): 1080x1920 @ 2.04x in 24", 60 Hz [External]
Display (EV2450): 1920x1080 @ 1.15x in 24", 60 Hz [External]
WM: niri 26.04 (Wayland)
CPU: Intel(R) Core(TM) i5-8365U (8) @ 4.10 GHz
GPU: Intel UHD Graphics 620 @ 1.10 GHz [Integrated]


EDIT.: Screen updates when moving mouse on the vertical monitor. Maybe some script that moves a virtual mouse every fraction of a second on the vertical monitor could work? Or is there some command that can force redraws?

**Comment by Leo328** [comment]
Hi there,

I have a similar issue which I've spent over 6 hours trying to debug and fix on my machine - I have done everything I reasonably can to fix it via software and kept running into a brick wall.

For the sake of trying to save someone else a massive headache, this is what I tried.  In the end, I could not get DisplayLink to output flipped screens and had to revert to connecting the display directly via DisplayPort instead, bypassing DisplayLink entirely.

**I tried pretty much all the configuration flags that were possible to toggle and nothing got it to work properly**

Here is my report:

## Summary

I am seeing a reproducible issue with rotated DisplayLink/EVDI outputs on Linux when NVIDIA is the primary Xorg GPU.

A DisplayLink-connected monitor works perfectly in normal landscape orientation. As soon as any rotation/transform is applied (`inverted`, and also reflection in testing), the display itself remains active and the cursor/background may remain visible, but application windows can no longer be displayed normally on that monitor.

The strongest A/B test is that I moved the exact same physical monitor from a DisplayLink output to the dock's native DisplayPort passthrough output. It then appeared directly as an NVIDIA `DP-0` output.

Rotation immediately worked perfectly there, including application windows.

This strongly suggests the problem is specifically in the NVIDIA -> EVDI transformed-output/presentation path rather than Cinnamon's rotation implementation, the monitor, or the physical display orientation.

---

## Hardware

- GPU 1 / primary Xorg GPU: NVIDIA RTX 5060 Ti 16 GB
- GPU 2: AMD Radeon Pro V620 32 GB
  - secondary/offload GPU
- Dock:
  - Kensington SD5900T / SD5910T / SD5920T EQ Thunderbolt 4 Quad 4K Hybrid Dock
  - DisplayLink chipset identifies as:
    `17e9:6000 DisplayLink DL-DOCK`
- Main monitor:
  - 3840x2160
  - directly attached to NVIDIA 5060 Ti via HDMI
- Two Dell P2219H 1920x1080 monitors connected through the dock
- One Dell needs to operate upside-down (`xrandr --rotate inverted`)

The dock is connected over Thunderbolt/USB4 at 40 Gb/s.

---

## Software

- Linux Mint 22.3
- Cinnamon 6.6.9
- X11
- Xorg 21.1.12
- DisplayLink driver 6.3.0-48
- EVDI 1.15.0

Tested kernels:

- 7.0.0-28-generic
- 6.17.0-35-generic

Tested NVIDIA drivers:

- 580-open
- 595-open
- Current test driver: 595.84

`nvidia_drm` KMS is enabled.

---

## Normal working configuration

With both DisplayLink monitors unrotated:

    HDMI-0      3840x2160+0+0       normal
    DVI-I-2-1   1920x1080+3840+0    normal
    DVI-I-3-2   1920x1080+3840+1080 normal

Both DisplayLink monitors render correctly and application windows can be dragged between all displays.

For example:

    xrandr \
      --output DVI-I-2-1 --mode 1920x1080 --pos 3840x0 --rotate normal \
      --output DVI-I-3-2 --mode 1920x1080 --pos 3840x1080 --rotate normal

works correctly.

---

## Reproduction

Starting from the fully working configuration above:

    xrandr --output DVI-I-2-1 --rotate inverted

The monitor visibly rotates, but application-window presentation on that monitor stops working.

Symptoms have included:

- monitor remains detected and enabled
- desktop/background can remain visible
- mouse cursor remains visible and responsive
- application windows cannot be dragged/rendered normally onto the rotated monitor
- switching the output back to `normal` restores normal operation

The second, unrotated DisplayLink monitor continues working.

This is highly reproducible.

---

## Other transforms

Reflection was also tested, e.g. using an XRandR reflection transform.

This caused essentially the same failure: the transformed DisplayLink output stopped presenting application windows correctly.

So the issue does not appear to be limited specifically to the `inverted` orientation.

---

## Cinnamon/Muffin test

To exclude an XRandR/Cinnamon state mismatch, rotation was also applied through:

    org.cinnamon.Muffin.DisplayConfig

The Muffin state correctly reported transform `2`, and XRandR also reported the monitor as `inverted`.

Application windows still failed to render on the DisplayLink monitor.

Therefore this does not appear to be caused by Cinnamon simply having stale logical-monitor geometry.

Restarting Cinnamon restored window rendering, but also reset the problematic rotation.

---

## DisplayLink/Xorg workarounds tested

I tested an EVDI OutputClass containing combinations of:

    Option "PageFlip" "off"
    Option "ShadowFB" "true"
    Option "AccelMethod" "none"
    Option "SWCursor" "on"

Results:

### PageFlip off

No fix.

Normal orientation remained smooth, but the rotated monitor still failed to display application windows.

### ShadowFB true

No fix.

### AccelMethod none

No fix.

### SWCursor on

This produced a very interesting result.

With:

    Option "SWCursor" "on"

application windows DID start rendering on the rotated DisplayLink monitor.

However, overall display performance became extremely choppy and severe graphical artifacts appeared across other displays/windows.

The system was not usable in this state.

This was reproducible both with SWCursor alone and with the larger collection of DisplayLink Xorg workarounds.

This seems potentially significant because SWCursor changes the otherwise broken behavior of the rotated framebuffer.

---

## NVIDIA workarounds tested

### AllowFlipping

Tested dynamically:

    nvidia-settings -a AllowFlipping=0

No improvement.

Rotation still resulted in the DisplayLink monitor failing to display application windows.

Restored afterwards with:

    nvidia-settings -a AllowFlipping=1

### DamageEvents

Tested NVIDIA Xorg option:

    Option "DamageEvents" "false"

No fix and it caused additional display initialization problems, so it was reverted.

---

## Different rendering/session paths tested

### Cinnamon software rendering

Tested Cinnamon's software-rendering session.

Same problem, if anything worse:

- rotated DisplayLink screen
- cursor visible
- desktop/application rendering failed

Therefore this does not appear to depend on Cinnamon's normal GPU-accelerated compositor.

### Wayland

Cinnamon Wayland Experimental was also tested.

Unlike X11, application windows could render on the DisplayLink outputs, including transformed displays, but both DisplayLink displays became extremely laggy and were not practically usable.

---

## Different kernels

Tested:

- Linux 7.0.0-28
- Linux 6.17.0-35

Behavior was identical.

---

## Different NVIDIA drivers

Tested:

- NVIDIA 580-open
- NVIDIA 595-open / 595.84

Behavior was identical.

Therefore this does not appear to be a regression specific to NVIDIA 580 vs 595.

---

## EVDI initialization / DRM ordering

Originally EVDI could initialize before the physical GPUs.

Example:

    card0 -> evdi
    card1 -> NVIDIA
    card2 -> AMD

The EVDI configuration was corrected to:

    options evdi initial_device_count=4
    softdep evdi pre: amdgpu nvidia_drm

After reboot the ordering became:

    card1 -> AMD
    card2 -> NVIDIA
    card3 -> evdi.0
    card4 -> evdi.1
    card5 -> evdi.2
    card6 -> evdi.3

Xorg then had all four EVDI providers available from startup.

This improved/cleaned up EVDI/Xorg initialization, but DID NOT change the rotation problem.

---

## Provider topology

Working Xorg provider topology is approximately:

    Provider 0: NVIDIA-0
        Source Output

    Provider 1-4: modesetting
        Sink Output
        EVDI devices

    Provider 5: AMD Radeon Pro V620
        Sink Output, Source Offload

So NVIDIA is the Xorg source driving the EVDI sink outputs.

---

## AMD-primary experiment

I also attempted making the AMD V620 the primary Xorg rendering device and using NVIDIA as an output sink, to determine whether eliminating NVIDIA -> EVDI presentation would resolve the problem.

The resulting X configuration failed to bring up the graphical desktop correctly, so this test could not be completed and was reverted.

---

## DisplayLink documented monitor-position workaround

The workaround of placing the monitor at the extreme left/right of the desktop before applying rotation was tested.

Normal orientation worked at the edge.

Applying rotation immediately reproduced the same broken-window behavior.

---

## Relevant NVIDIA/EVDI kernel errors

During testing I repeatedly observed errors similar to:

    [nvidia-drm] Failed to get memory pages for NvKmsKapiMemory

followed by:

    evdi_gem_mmap: Don't allow imported objects to be mapped: owner: nvidia_drm

These appeared multiple times during NVIDIA/EVDI operation.

I do not know whether these are directly responsible for the rotation issue, but they may be relevant to buffer sharing between `nvidia_drm` and EVDI.

---

## Critical A/B test: native DisplayPort

Finally I moved the exact same top Dell monitor away from DisplayLink.

The dock has one DisplayPort output that uses native DP passthrough rather than DisplayLink.

The topology became:

    NVIDIA RTX 5060 Ti
        |
        DP output
        |
    motherboard DP-IN
        |
    Thunderbolt
        |
    Kensington dock native DP
        |
    Dell P2219H

Linux then detected this display directly on NVIDIA as:

    DP-0 connected 1920x1080

I applied:

    xrandr --output DP-0 --rotate inverted

It worked perfectly.

- display rotated correctly
- desktop rendered correctly
- application windows rendered correctly
- windows could be dragged onto it normally
- no lag or artifacts

The bottom Dell remains connected through DisplayLink/EVDI and works normally as long as it is not transformed.

Therefore the current workaround is:

    HDMI-0
        -> native NVIDIA
        -> main 4K display

    DP-0
        -> native NVIDIA via Thunderbolt DP passthrough
        -> top Dell
        -> inverted
        -> works

    DVI-I-3-2
        -> DisplayLink/EVDI
        -> bottom Dell
        -> normal
        -> works

---

## Conclusion / suspected problem

Based on the tests above, the problem appears to be specifically related to framebuffer presentation/update when a transform is applied to an EVDI sink being driven from an NVIDIA Xorg source.

The evidence is:

    Same monitor
    Same cable
    Same system
    Same Cinnamon session
    Same NVIDIA GPU

    DisplayLink/EVDI + normal
        -> works

    DisplayLink/EVDI + rotation
        -> application presentation fails

    Native NVIDIA + normal
        -> works

    Native NVIDIA + identical rotation
        -> works perfectly

**Also interesting is that enabling EVDI `SWCursor` changes the behavior enough that application windows begin appearing on the transformed EVDI output, although with severe performance/artifact problems.**

**That may indicate an interaction involving damage tracking, buffer mapping/presentation, or NVIDIA/EVDI framebuffer sharing rather than a failure of rotation geometry itself.**


With Linux massively increasing in popularity, could someone at DisplayLink/Synaptics please fix this issue?

My setup is definitely not common, but it does uncover some pretty core issues that need to be resolved.

My entire ChatGPT conversation debugging this is here:
[Linux Mint Multi Monitor setup issue](https://chatgpt.com/share/6a765725-3778-83e8-91f8-614d610bd2a5 )
